### ISSUE TITLE
Specify individual GPU for `train_vae`?

### ISSUE THREAD

**Issue by olibclarke** [zhonge/cryodrgn#257]
Hi,

I might be (likely am!) missing something very obvious, but how do I specify a single GPU on which to run `train_vae`, is there a flag for this? I see that it can be done by setting an environment variable, `$CUDA_VISIBLE_DEVICES`, is this the only/preferred method?

Cheers
Oli

**Comment by Guillawme** [comment]
Yes, setting `$CUDA_VISIBLE_DEVICES` before the `cryodrgn` command works perfectly well (at least it did when I used it; I haven't run cryodrgn recently).

**Comment by zhonge** [comment]
Yes, the cleanest way is to set the environment variable `CUDA_VISIBLE_DEVICES`:

```
CUDA_VISIBLE_DEVICES=0 cryodrgn train_vae ...
```

I believe it is the recommended way to select the desired GPU for training pytorch models.

**Comment by michal-g** [comment]
Would there be any interest in making this a flag to help users, and then doing something like `os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(args.cuda_devices)` within the script?

Many users may find doing the extra command-line step frustrating, and there might be pitfalls too (like not using `export` to set it as an environment variable).

**Comment by Guillawme** [comment]
I guess this could simplify the user interface a bit, but the default value (when the new command-line option you are suggesting is not set) should be considered carefully. Should it be "the first visible GPU" or "all visible GPUs"? I think it depends on how one runs cryoDRGN. If I am running the command on a local workstation, I would expect it to use the first visible GPU by default, not saturate the workstation right away. But if I am running it on a cluster and the submission script booked more than one GPU, I would want cryoDRGN to use them all (I guess in this case it is fine to require setting the option explicitly, since one would have to write a script in the first place anyway to submit a batch job).

**Comment by michal-g** [comment]
That makes sense — we already have the `--multigpu` flag for `train_nn` and `train_vae` (some more work needs to happen for it to be extended to `abinit_homo` and `abinit_het`) so the default behaviour for these commands is already "use the first visible GPU" and then "all visible GPUs" with the flag. So it seems to me that the default values for `args.cuda_device` could be decided by just looking at `args.multigpu`, and we should continue preventing users from using all GPUs unless they specify otherwise!
